### ISSUE TITLE
Clean up reek_of/smell_of matchers

### ISSUE THREAD

**Issue by mvz** [troessner/reek#319]
We have both reek_of and smell_of matchers. This seems silly. Perhaps we can make reek_of smarter so it doesn't check all smells but just the one being tested. This should be possible now that SmellRepository can be initialized with the smells to test.

**Comment by mvz** [comment]
I think we should pick this up for Reek 2.0 as well, since it might involve a breaking change to `#reek_of`.

**Comment by troessner** [comment]
Yes, makes total sense to me.

**Comment by troessner** [comment]
Started to work on this in #371 (which is highly wip).

**Comment by troessner** [comment]
Regarding:

>> We have both reek_of and smell_of matchers. This seems silly. 

Let's compare "smell_of" with "reek_of":

**smell_of:**

spec/reek/smells/long_yield_list_spec.rb:      

```Ruby
expect(src).to smell_of(LongYieldList, count: 4)
```

spec/reek/smells/control_parameter_spec.rb:

```Ruby
expect(src).to smell_of(ControlParameter)
expect(src).to smell_of(ControlParameter, name: 'arg')
```

spec/reek/smells/data_clump_spec.rb:    

```Ruby
expect(src).to smell_of(DataClump,
                            parameters: %w(p1 p2)

expect(src).to smell_of(DataClump,
                            count: 3,
                            parameters: ['pa', 'pb', 'pc'])
```
**reek_of:**

spec/reek/core/module_context_spec.rb:

```Ruby
    expect('
      module Fred
        def simple(x) x + 1; end
      end
    ').to reek_of(:UncommunicativeParameterName, /x/, /simple/)
```

spec/reek/smells/irresponsible_module_spec.rb:


```Ruby
expect(src).to reek_of(:IrresponsibleModule)
```

spec/reek/smells/uncommunicative_variable_name_spec.rb:

```Ruby
expect(src).to reek_of(:UncommunicativeVariableName, /@x/, /Thing/, /variable name/)
```

**Things that I notice when looking at "smell_of":**

- it's not very caller-friendly - it expects a class, and if we pass a wrong constant here the method doesn't catch and handle this in a friendly way but rather the method call itself will raise.
- smell_of is a very sophisticated matcher. It allows for exact matching 
- Nobody would use this matcher as a client, at least not for precise matching. Nobody. This one makes only sense for us, the `reek` contributors, to test `reek` itself. 


**Things that I notice when looking at "reek_of":**

- it takes a symbol, so at least in theory it's more caller-friendly (or rather: can be)
- it doesn't offer the same level of complexity like "smell_of"
- overall it seems more suited for clients with calls like:
```Ruby
expect(src).to reek_of(:IrresponsibleModule)
```
- but something like
```Ruby
expect(src).to reek_of(:UncommunicativeVariableName, /@x/, /Thing/, /variable name/)
```
is completely over the top. I have a hard time imagining somebody testing on this level of granularity. If you would test like that you would mirror the implementation. [1]

**Suggestion**

I'd merge the two methods like this:

- remove "smell_of" completely
- remove the second parameter from reek_of, the "patterns" because of [1]
- add the second parameter from "smell_of" to "reek_of"

So in the end I'd be able to call "reek_of" like this:

```Ruby
expect(src).to reek_of(DataClump,
                            count: 3,
                            parameters: ['pa', 'pb', 'pc'])

```

but also just like this:

```Ruby
expect(src).to reek_of(:IrresponsibleModule)
```

What do you guys think about that?

**Comment by troessner** [comment]
pinging especially @mvz since he made the initial suggestion.:)

**Comment by troessner** [comment]
Gonna start today with this, quick feedback would be great.;)

**Comment by mvz** [comment]
Yes, I think this is a good idea as well. One detail: Current `smell_of` allows passing an arbitrary number of hashes, and it will check that it smells as many times with those exact parameters. I think this is overkill, and you can just call `reek_of` with a hash several times. If we need to ensure the number of smells, we can add `.exactly(3).times`, or however that is done in RSpec.

**Comment by troessner** [comment]
>>  Current smell_of allows passing an arbitrary number of hashes, and it will check that it smells as many times with those exact parameters. I think this is overkill, and you can just call reek_of with a hash several times. If we need to ensure the number of smells, we can add .exactly(3).times, or however that is done in RSpec.

Yes, couldn't agree more.

**Comment by troessner** [comment]
merged #371
